### PR TITLE
First proof of concept - not for merging

### PR DESCRIPTION
**Pull request by marktheunissen**
This is my idea for how plugins can provide mixins - e.g. you can have methods in the MySQL plugin that work with MySQL on Linux, Ubuntu, etc. and those methods are added to the Environment object at runtime for use by the whole system. 

What do you think of this approach? This PR is just for comment, showing working code moving userExists into ApacheHTTPD (we won't actually do this)

### CODE REVIEW

**Comment by rupl** [comment]
This is awesome, merging immediately.

File: controller/src/TrainingWheels/Plugin/ApacheHTTPD/ApacheHTTPD.php [hunk 1]
```diff
@@ -10,6 +11,22 @@ public function __construct() {
     $this->ansible_play = __DIR__ . '/ansible/apachehttpd.yml';
   }
 
+  public function getEnvMixins($type) {
```
Comment (marktheunissen) [inline controller/src/TrainingWheels/Plugin/ApacheHTTPD/ApacheHTTPD.php]: this is the magic where we provide anonymous functions.

**Comment by marktheunissen** [comment]
:poop:

**Comment by marktheunissen** [comment]
This is also the first of a series of many refactorings of this particular part of the code.

**Comment by elliotttf** [comment]
This does feel a little like black-magic but that seems like a fair tradeoff if:
1. Our examples are well documented
2. It does indeed give us the flexibility that it looks like it would.

:+1:

**Comment by marktheunissen** [comment]
Yeah, I think it maximizes flexibility, because no part of the system has to care _which_ plugin provides a method on the env object. A lot of this design for flexibility is speculative, but it's not too confusing, the API is quite easy to understand I think. Can always make it more explicit down the line if it becomes a mess.

**Pull request closed without merging**